### PR TITLE
feat: implement urlpattern

### PR DESCRIPTION
**Pull request by crowlKats**

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+      return;
+    }
+
+    let mut regexp_value: &str = if regexp_or_wildcard_token.is_none() {
+      &self.segment_wildcard_regexp
+    } else if regexp_or_wildcard_token.unwrap().kind == TokenType::Asterisk {
+      FULL_WILDCARD_REGEXP_VALUE
+    } else {
+      &regexp_or_wildcard_token.unwrap().value
+    };
```
Comment (lucacasonato) [inline src/parser.rs:180]: This could be a match statement :-)

File: src/parser.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        } else {
+          regexp_or_wildcard_token.value.to_owned()
+        }
+      }
+    };
+
+    let mut kind = PartType::Regexp;
+    if regexp_value == self.segment_wildcard_regexp {
+      kind = PartType::SegmentWildcard;
+      regexp_value = String::new();
```
Comment (sno2) [inline src/parser.rs:206]: Would [`String.clear()`](https://doc.rust-lang.org/std/string/struct.String.html#method.clear) be better for resetting the contents here and in other places like this?  Although, I'm not sure if the part of it not resetting the capacity could be considered bad.

```suggestion
      regexp_value.clear();
```

File: src/tokenizer.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    tokenizer.index,
+    tokenizer.index,
+  );
+  Ok(tokenizer.token_list)
+}
+
+// Ref: https://wicg.github.io/urlpattern/#is-a-valid-name-code-point
+#[inline]
+fn is_valid_name_codepoint(_code_point: char, _first: bool) -> bool {
+  todo!("issue: there is a unicode_xid crate, but sadly that's xid, whereas the spec asks for id")
```
Comment (littledivy) [inline src/tokenizer.rs:292]: `unic_ucd_ident::is_id_start` with `features = ["id"]`

File: src/constructor_parser.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    if self.is_non_special_pattern_char(self.token_index, "?") {
+      return true;
+    }
+    if self.token_list[self.token_index].value != "?" {
+      return false;
+    }
+    let previous_index = self.token_index - 1;
+    #[allow(clippy::absurd_extreme_comparisons)]
+    if previous_index < 0 {
+      // TODO: can self.token_index ever be negative?
```
Comment (lucacasonato) [inline src/constructor_parser.rs:68]: not if you are using a `usize` (unsigned)

Comment (crowlKats) [reply src/constructor_parser.rs:68]: I am aware, just thinking in terms of spec. frm what i can see from the spec though, i dont see that being the case, but i might have missed something

File: src/canonicalize_and_process.rs [hunk 5]
```diff
@@ -35,40 +37,52 @@ pub fn canonicalize_hostname(value: &str) -> Result<String, ParseError> {
...
-  todo!()
+  if let Some(protocol) = protocol {
+    let mut url =
+      url::Url::parse(&format!("{}://dummy.test", protocol)).unwrap(); // TODO: dont unwrap, instead ParseError
+    url::quirks::set_port(&mut url, value).unwrap(); // TODO: dont unwrap, instead ParseError
+    Ok(url::quirks::port(&url).to_string())
+  } else {
+    // If no protocol is given, the url can not have a username/password/port, so
+    // we can always return the empty string
+    Ok("".to_string())
```
Comment (sno2) [inline src/canonicalize_and_process.rs:52]: I think you should just use `String::new()`, no need to create an `&str` then convert it.

**Comment by lucacasonato** [comment]
Stuff left to do:

- [ ] Add unit tests for the tokenizer
- [ ] Add unit tests for the constructor string parser
- [ ] Add unit tests for the pattern parser
- [ ] Remove `SomeRandomOtherError`
- [ ] Rename `ParseError` enum to just `Error` (public name will be `urlpattern::Error`
- [ ] Add public API docs, and README

**Review by lucacasonato — APPROVED** [review]
Awesome work @crowlKats! Let's land this, and continue iterating on main.

**Pull request merged**